agent/taskresource/cgroup/cgroup.go

```diff
@@ -0,0 +1,212 @@
+}
+
+// Create creates cgroup root for the task
+func (c *CgroupResource) Create(task *api.Task) error {
+	err := c.setupTaskCgroup(task)
+	if err != nil {
+		seelog.Criticalf("Unable to setup platform resources: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (c *CgroupResource) setupTaskCgroup(task *api.Task) error {
+	cgroupRoot := c.CgroupRoot
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	_, err = c.control.Create(&cgroupSpec)
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to create cgroup at %s for task: %s", cgroupRoot, task.Arn)
+	}
+
+	// enabling cgroup memory hierarchy by doing 'echo 1 > memory.use_hierarchy'
+	memoryHierarchyPath := filepath.Join(c.CgroupMountPath, memorySubsystem, cgroupRoot, memoryUseHierarchy)
+	err = c.ioutil.WriteFile(memoryHierarchyPath, enableMemoryHierarchy, rootReadOnlyPermissions)
+	if err != nil {
+		return errors.Wrap(err, "resource: setup cgroup: unable to set use hierarchy flag")
+	}
+
+	return nil
+}
+
+// Cleanup removes the cgroup root created for the task
+func (c *CgroupResource) Cleanup() error {
+	err := c.control.Remove(c.CgroupRoot)
+	// Explicitly handle cgroup deleted error
+	if err != nil {
+		if err == cgroups.ErrCgroupDeleted {
+			seelog.Warnf("Cgroup at %s has already been removed: %v", c.CgroupRoot, err)
+			return nil
+		}
+		return errors.Wrapf(err, "resource: cleanup cgroup: unable to remove cgroup at %s", c.CgroupRoot)
+	}
+	return nil
+}
+
+// cgroupResourceJSON duplicates CgroupResource fields, only for marshalling and unmarshalling purposes
+type cgroupResourceJSON struct {
+	CgroupRoot      string        `json:"CgroupRoot"`
+	CgroupMountPath string        `json:"CgroupMountPath"`
+	CreatedAt       time.Time     `json:",omitempty"`
+	DesiredStatus   *CgroupStatus `json:"DesiredStatus"`
+	KnownStatus     *CgroupStatus `json:"KnownStatus"`
+}
+
+// MarshalJSON marshals CgroupResource object using duplicate struct CgroupResourceJSON
+func (c *CgroupResource) MarshalJSON() ([]byte, error) {
+	return json.Marshal(cgroupResourceJSON{
+		c.CgroupRoot,
+		c.CgroupMountPath,
+		c.GetCreatedAt(),
+		func() *CgroupStatus { desiredState := c.GetDesiredStatus(); return &desiredState }(),
+		func() *CgroupStatus { knownState := c.GetKnownStatus(); return &knownState }(),
+	})
+}
+
+// UnmarshalJSON unmarshals CgroupResource object using duplicate struct CgroupResourceJSON
+func (c *CgroupResource) UnmarshalJSON(b []byte) error {
+	temp := &cgroupResourceJSON{}
+
+	if err := json.Unmarshal(b, &temp); err != nil {
```

Comment:
Doesn't this mean that you're passing a pointer to a pointer? `json.Unmarshal(b, temp)` should be good enough i think